src/prefect/results.py

```diff
@@ -43,3 +56,201 @@ async def _retrieve_result(state):
+        persist_result = (
+            task.persist_result
+            if task.persist_result is not None
+            else
+            # !! Tasks persist their result by default if their parent flow uses a
+            #    feature that requires it or the task uses a feature that requires it
+            (
+                flow_features_require_child_result_persistence(ctx.flow)
+                or task_features_require_result_persistence(task)
+            )
+        )
+
+        return await cls.from_settings(
+            result_storage=result_storage,
+            result_serializer=result_serializer,
+            persist_result=persist_result,
+            client=client,
+        )
+
+    @classmethod
+    @inject_client
+    async def from_settings(
+        cls: Type[Self],
+        result_storage: ResultStorage,
+        result_serializer: ResultSerializer,
+        persist_result: bool,
+        client: OrionClient,
+    ) -> Self:
+        storage_block_id, storage_block = await cls.resolve_storage_block(
+            result_storage, client=client
+        )
+        serializer = cls.resolve_serializer(result_serializer)
+
+        return cls(
+            storage_block=storage_block,
+            storage_block_id=storage_block_id,
+            serializer=serializer,
+            persist_result=persist_result,
+        )
+
+    @staticmethod
+    async def resolve_storage_block(
+        result_storage: ResultStorage, client: OrionClient
+    ) -> Tuple[uuid.UUID, WritableFileSystem]:
+        """
+        Resolve one of the valid `ResultStorage` input types into a saved block
+        document id and an instance of the block.
+        """
+        if isinstance(result_storage, uuid.UUID):
+            storage_block_id = result_storage
+            document = await client.read_block_document(storage_block_id)
+            storage_block = Block._from_block_document(document)
+        elif isinstance(result_storage, Block):
+            storage_block = result_storage
+            storage_block_id = (
+                # Avoid saving the block if it already has an identifier assigned
+                storage_block._block_document_id
+                # TODO: Overwrite is true to avoid issues where the save collides with
+                #       a previously saved document with a matching hash
+                or await storage_block._save(is_anonymous=True, overwrite=True)
+            )
+        elif isinstance(result_storage, str):
+            storage_block = await Block.load(result_storage, client=client)
+            storage_block_id = storage_block._block_document_id
+            assert storage_block_id is not None, "Loaded storage blocks must have ids"
+        else:
+            raise TypeError(
+                "Result storage must be one of the following types: 'UUID', 'Block', "
+                f"'str'. Got unsupported type {type(result_storage).__name__!r}."
+            )
+
+        return storage_block_id, storage_block
```

Comment:
Do we need to check if the block has file read/write capability?

Reply:
Yeah I think we need a utility to do so. It'd be great to check earlier (e.g. at flow declaration) rather than at runtime (when this object is created).